### DISCUSSION TITLE
How do I combine multiple `TryAsync` results?

### DISCUSSION THREAD

**Discussion by MrYossu** [louthy/language-ext#961] [Q&A]
Suppose I have the following methods...

```c#
TryAsync<File> GetFolder(string folderId)
TryAsync<string> CreateFolder(string folderName, string parentFolderId)
TryAsync<string> UploadFile(Stream file, string fileName, string mimeType, string folderId)
```

...where `File` is a Google type for files/folders in a Google Drive.

I can call each of these individually, no problem. However, I sometimes want to call more than one, say to get the `File` object for a specific folder, then create a new subfolder and upload a file to it. I know I can do this as follows (air code, so please ignore any typos)...

```c#
(await GetFolder("123"))
  .Match(async folder => {
    (await CreateFolder("New folder", folder.Id))
      .Match(async newFolder => {
        (await UploadFile(stream, "New file name.txt", "text/text", newFolder.Id))
          .Match(fileId => /* do whatever with the uploaded file's Id */, ex => /* report exception */);
      }, ex => /* report exception */);
  }, ex => /* report exception */);
```

As you can see, this is very painful. I am sure that you are supposed to be able to chain monads together, I think using `Bind`, so you end up with something more like this (again, air code)...

```c#
(await GetFolder("123"))
  .Bind(folder => CreateFolder("New folder", folder.Id))
  .Bind(newFolder => UploadFile(stream, "New file name.txt", "text/text", newFolder.Id))
  .Match(fileId => /* do whatever with the uploaded file's Id */, ex => /* report exception */);
```

However, I can't get any code to compile like this.

Anyone able to advise how I can improve the nested monads I have? Thanks

**Comment by Malgefor** [comment]
You bind the `TryAsync` operations together like this:

```csharp
            var bindWayResult = await GetFolder("123")
                .Bind(folder => CreateFolder("New folder", folder))
                .Bind(createdFolderResult => UploadFile(Stream.Null, "New file name.txt", "text/text", createdFolderResult))
                .Match(successValue => successValue, exception => $"Exception: {exception.Message}");
            
            var linqWayOperation =
                from folder in GetFolder("123")
                from createdFolderResult in CreateFolder("New folder", folder)
                from uploadFileResult in UploadFile(Stream.Null, "New file name.txt", "text/text", createdFolderResult)
                select uploadFileResult;

            var linqWayOperationResult = await linqWayOperation
                .Match(successValue => successValue, exception => $"Exception: {exception.Message}");
```

You can also use `Aff` instead of `TryAsync`. I think this is the way going forward since release 4 of Language.Ext.
[Here is more documentation](https://github.com/louthy/language-ext/wiki/How-to-deal-with-side-effects#aff-and-eff-monad)

**Comment by MrYossu** [comment]
@Malgefor Thanks for the reply, but unfortunately, it didn't work. I copied your first example (as I prefer the method syntax), and changed the second parameter to the `CreateFolder` method to `folder.Id`, as the `CreateFolder` method  returns a `File` object, but `CreateFolder` expects a string (the folder Id) as its second parameter. This gave...

```c#
var bindWayResult = await GoogleDriveHelper.GetFolder("123")
  .Bind(folder => GoogleDriveHelper.CreateFolder("New folder", folder.Id))
  .Bind(createdFolderResult => GoogleDriveHelper.UploadFile(Stream.Null, "New file name.txt", "text/text", createdFolderResult))
  .Match(successValue => successValue, exception => $"Exception: {exception.Message}");
```

However, that gave me a compiler error on the first call to `Bind`...

_The type arguments for method 'TryAsyncExtensions.Bind<A, B>(TryAsync<A>, Func<A, TryAsync<B>>)' cannot be inferred from the usage. Try specifying the type arguments explicitly._

Hovering over the `folder` parameter at the beginning of the lambda shows it to be of the `File` type

Any ideas? Thanks again.

**Comment by Malgefor** [comment]
When I copy your code it compiles just fine. I think the problem is something else than the `Bind`-method.

- Are you sure that you are using the same `File`-type everywhere? Not accidentally a mix between the `Google`-type and `System.IO.File`?
- Are the method signatures still the same as in your original question? What is the return type of `CreateFolder` in your first `Bind`-call?

**Comment by MrYossu** [comment]
@Malgefor Duh, you got it! I had forgotten that when I first wrote these methods, I went through a few iterations of using the wrong monads. The previous iteration returned an `Either<Exception, T>`. I then realised that `TryAsync` catches and handles the exception for you, reducing the amount of code you need. I converted all the methods except one to use `TryAsync`, but had trouble with `CreateFolder` as I couldn't work out how to write it. I ended up leaving it returning an `Either<Exception, T>`, but copied the wrong bit of code into my question.

With a dummy body for `CreateFolder`, the code you showed compiles fine. All I need to do now is work out how to convert that last method to use `TryAsync`. 

Thanks again. If I get stuck on converting, I'll open a new issue.

**Comment by MrYossu** [comment]
@Malgefor Got stuck with that last method. I opened [a new issue](https://github.com/louthy/language-ext/issues/956), so if there's any chance you could take a look, I'd really appreciate it. 

Thanks again
